docs/rfcs/0006-top-sites-feature.md

```diff
@@ -0,0 +1,125 @@
+---
+layout: page
+title: Introduce a Top Sites Feature
+permalink: /rfc/0006-top-sites-feature
+---
+
+* Start date: 2020-07-30
+* RFC PR: [#7931](https://github.com/mozilla-mobile/android-components/pull/7931)
+
+## Summary
+
+Create a TopSitesFeature that abstracts out the co-ordination between the multiple data storages.
+
+## Motivation
+
+A large part of the top sites implementation is spread across A-C and Fenix. There is a lot of co-ordination between the History API and the current Top Sites API to manage the final state of what sites should be displayed that can exist as a single implementation.
+
+## Guide-level explanation
+
+To begin, we should rename the `TopSitesStorage`  to `PinnedSitesStorage` and create a `TopSitesFeature` that would orchestrate the different storages and present them to the app.
+
+We can then follow the architecture of existing features by using the presenter/interactor/view model, and introduce a `TopSitesStorage` which abstract out the complexity of the multiple data sources from `PinnedSitesStorage` and `PlacesHistoryStorage`.
+
+To allow the ability for adding a top site from different parts of the app (e.g. context menus), we introduce the `TopSitesUseCases` that acts on the storage directly. If the `TopSitesFeature` is started, it will then be notified by the `Observer` to update the UI.
+
+```kotlin
+
+/**
+ * Implemented by the application for displaying onto the UI.
+ */
+interface TopSitesView {
+  /**
+   * Update the UI.
+   */
+  fun displaySites(sites: List<TopSite>)
+
+  interface Listener {
+    /**
+     * Invoked by the UI for us to add to our storage.
+     */
+    fun onAddTopSite(topSite: TopSite)
```

Comment:
Let's get the terminology right here. What does that mean, "addTopSite"? Do you actually mean "pin site"? I think so, since frecent sites are outside of our control, and default sites don't change.

Let's not mix-up the naming at the api level. We "get a list of top sites" (since it's a merge of three data sources), but we create "pinned sites".

Reply:
We kept `onAddTopSite` here since it exists on the API level and not the implementation we have (which is a pinned site). Using top sites instead of pinned sites conforms to your standard naming of the feature itself on desktop too.

Should we switch them up to have the API `onAddPinnedSite` everywhere?

Reply:
Having the API be in terms of adding a pinned site makes sense to me, since it makes it very clear what's actually happening. E.g. it's not at all clear what "addTopSite" would do for a feature which operates over three different sources of data.

You'll also be exposing pinned-sites storage to the app, right? Perhaps this isn't even necessary then? E.g. the application can query via the feature, but add via the storage. (and then re-query after adding).

Reply:
TODO

- [ ] Rename `onAddTopSite` to `onAddPinnedSite`

Reply:
In terms of naming, this looks a bit odd. A `TopSitesView` that has callbacks to `onAddPinnedSite`?

We can rename the view class to `PinnedSitesView`, but that as a chain reaction of needing to rename everything to "PinedSites". We will then also have a "DefaultPinedSitesStorage" that takes a `PinedSitesStorage` and `PlacesHistoryStorage`. 😃

Reply:
Let's punt on this renaming, because it sounds like we can get rid of the `PinedSitesStorage` when we migrate it over the history storage API and then we would have a straight-forward naming for what we call "top sites" everywhere.